vpr/src/route/heap_type.h

```diff
@@ -0,0 +1,162 @@
+// Interface to heap used for router optimization.
+//
+// Note: Objects used in instances of HeapInterface must always be allocated
+// and free'd using the HeapInterface::alloc and HeapInterface::free methods
+// of that instance.  Object pools are likely in use.
+//
+// As a general rule, any t_heap objects returned from this interface,
+// **must** be HeapInterface::free'd before destroying the HeapInterface
+// instance.  This ensure that no leaks are present in the users of the heap.
+// Violating this assumption may result in a assertion violation.
+class HeapInterface {
+  public:
+    virtual ~HeapInterface() {}
+
+    // Allocate a heap item.
+    //
+    // This transfers ownership of the t_heap object from HeapInterface to the
+    // caller.
+    virtual t_heap* alloc() = 0;
+
+    // Free a heap item.
+    //
+    // HeapInterface::free can be called on objects returned from either
+    // HeapInterface::alloc or HeapInterface::get_heap_head.
+    virtual void free(t_heap* hptr) = 0;
+
+    // Initializes heap storage based on the size of the device.
+    //
+    // Note: this method **must** be invoked at least once prior to the
+    // following methods being called:
+    //  - add_to_heap
+    //  - push_back
+    //  - get_heap_head
+    //  - is_empty_heap
+    //  - empty_heap
+    //  - build_heap
+    virtual void init_heap(const DeviceGrid& grid) = 0;
+
+    // Add t_heap to heap, preserving heap property.
+    //
+    // This transfers ownership of the t_heap object to HeapInterface from the
+    // called.
+    virtual void add_to_heap(t_heap* hptr) = 0;
+
+    // Add t_heap to heap, however does not preserve heap property.
+    //
+    // This is useful if multiple t_heap's are being added in bulk.  Once
+    // all t_heap's have been added, HeapInterface::build_heap can be invoked
+    // to restore the heap property in an efficient way.
+    //
+    // This transfers ownership of the t_heap object to HeapInterface from the
+    // called.
+    virtual void push_back(t_heap* const hptr) = 0;
+
+    // Restore the heap property.
+    //
+    // This is useful in conjunction with HeapInterface::push_back when adding
+    // multiple heap elements.
+    virtual void build_heap() = 0;
+
+    // Pop the head of the pop, and return it.
+    //
+    // This transfers ownership of the t_heap object from HeapInterface to the
+    // caller.
+    virtual t_heap* get_heap_head() = 0;
+
+    // Is the heap empty?
+    virtual bool is_empty_heap() const = 0;
+
+    // Is the heap valid?
+    virtual bool is_valid() const = 0;
+
+    // Empty all items from the heap.
+    virtual void empty_heap() = 0;
+
+    // marks all the heap entries consisting of sink_node, where it was
+    // reached via ipin_node, as invalid (open).  used only by the
+    // breadth_first router and even then only in rare circumstances.
+    virtual void invalidate_heap_entries(int sink_node, int ipin_node) = 0;
+
+    // Free all storage used by the heap.
```

Comment:
Explain this truly frees the storage (returned to program's free list) and should be called only when the entire heap is no longer needed.

Reply:
Done